skimage/segmentation/fast_draw.py

```diff
@@ -0,0 +1,278 @@
+            * If `return_full_prob` is True, array of floats of same shape as
+              `image`. in witch each pixel is assigned the probability to
+              belong to `target_label`.
+          """
+        ## 1- Checking if inputs are valide
+        if target_label not in np.unique(labels):
+            warn('The target label '+str(target_label)+ \
+                 ' does not match any label')
+        if (labels != 0).all():
+            warn('The segmentation is computed on the unlabeled area '
+                 '(labels == 0). No zero valued areas in labels were '
+                 'found. Returning provided labels.')
+            segm = labels == target_label
+            if self.return_full_prob:
+                return segm.astype(np.float)
+            else:
+                return segm
+        
+        ## 2- Create down-sampled (coarse) image size 
+        ## and compute the energy map
+        ds_labels = np.zeros(self.dim)
+        ds_entropyMap = 0
+        for i in np.unique(labels):
+            if i != 0:
+                # 2.1- Compute the coarse label image
+                y,x = np.where(labels == i)
+                ds_labels[np.int32(y*self.full_to_ds_ratio[0]),
+                          np.int32(x*self.full_to_ds_ratio[1])] = i
+                # 2.2- Compute the energy map
+                M = np.ones_like(ds_labels)
+                M[ds_labels == i] = 0
+                distMap = distance_transform_edt(M)
+                ds_entropyMap +=  distMap
+        
+        # 2.3- Normalize the energy map and compute the ROI
+        ds_entropyMap = ds_entropyMap / ds_entropyMap.max()
+        threshold = ds_entropyMap.mean() + k*ds_entropyMap.std()
+        self.ds_maskROI = self.ds_maskROI | (ds_entropyMap <= threshold)
+    
+        ## 3- Performe a corse RW segmentation on the down-sampled image
+        unlabeled = np.ravel_multi_index(np.where((ds_labels == 0) & \
+                        (self.ds_maskROI)), ds_labels.shape)
+        labeled = np.ravel_multi_index(np.where((ds_labels > 0) & \
+                        (self.ds_maskROI)), ds_labels.shape)
+        # 3.1- Preparing the right handside of the equation BT xs
+        B = self.ds_L[unlabeled][:, labeled]
+        mask = ds_labels.flatten()[labeled] == target_label
+        fs = csr_matrix(mask)
+        fs = fs.transpose()
+        rhs = B * fs
+        # 3.2- Preparing the left handside of the equation Lu
+        Lu = self.ds_L[unlabeled][:, unlabeled]
+        # 3.3- Solve the linear equation Lu xu = -BT xs
+        probability = cg_solver(Lu, -rhs.todense(), tol=1e-3, maxiter=120)[0]
+
+        ds_proba = np.zeros_like(ds_labels, dtype=np.float32)
+        ds_proba[(ds_labels == 0) & (self.ds_maskROI)] = probability
+        ds_proba[(ds_labels == target_label) & (self.ds_maskROI)] = 1   
+        
+        # 3.4- Compute the corse segmentation result and the refinement region
+        #      around the corse result
```

Comment:
As mentioned in my general comment, this function is way too long. Your section titles are good candidates for where to break it up into component functions.